Spring boot security

Someboty opened the pull request:
Added User entity, Added Role entity, Added the required DTOs and mapper, Added AuthenticationController with the following methods: public UserLoginResponseDto login(UserLoginRequestDto request), public UserResponseDto register(UserRegistrationRequestDto request) throws RegistrationException,
Added FieldMatch annotation for checking that password and repeatPassword field in the UserRegistrationRequestDto are the same, Added the JWT support to the application

Someboty committed:
transport commit
Added User model, user dto, removed book models in separate package

Someboty committed:
Added User entity, Added Role entity, Added the required DTOs and mapper, Added AuthenticationController with the following methods: public UserLoginResponseDto login(UserLoginRequestDto request), public UserResponseDto register(UserRegistrationRequestDto request) throws RegistrationException,
Added FieldMatch annotation for checking that password and repeatPassword field in the UserRegistrationRequestDto are the same, Added the JWT support to the application

JeniaSan commented on src/main/java/com/bookshop/controller/AuthController.java:
```diff
@@ -0,0 +1,135 @@
+@RequiredArgsConstructor
+@RestController
+public class AuthController {
+    private static final String CODE_200 = "200";
+    private static final String TOKEN_GENERATED_DESCRIPTION =
+            "JWT token was generated successfully";
+    private static final String REGISTRATION_SUCCESSFUL =
+            "registration was successful";
+    private static final String CODE_204 = "204";
+    private static final String DELETED_USER_DESCRIPTION = "User deleted successfully";
+    private static final String ROLE_SET = "role was set to the user";
+    private static final String CODE_401 = "401";
+    private static final String CODE_401_DESCRIPTION =
+            "User should be authenticated to do that operation";
+    private static final String CODE_403 = "403";
+    private static final String CODE_403_DESCRIPTION =
+            "Only users with role \"ADMIN\" can do such operation";
+    private static final String CODE_500 = "500";
+    private static final String CODE_500_DESCRIPTION = "Internal server error";
+    private static final String INCORRECT_DATA =
+            """
+                {
+                    "timestamp": "*time of event*",
+                    "status": "INTERNAL_SERVER_ERROR",
+                    "errors": [
+                        "Incorrect user data"
+                    ]
+                }
+            """;
```
Do you really need all this code? Remember "KISS"

Someboty replied:
I thought it will look simple and nice. Codes, indeed, more for esthetic purpose, but example of exceptions are used in swagger, so I thought is useful for someone, who will read documentation. But you are not the first one, who propose to remove it, maybe I really should.

JeniaSan commented on src/main/java/com/bookshop/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+                        () -> new RegistrationException("Can't find user by email " + email));
+    }
+
+    private String prepareEmail(String query) {
+        String[] values = query.split(":");
+        return values[1].substring(1, values[1].substring(1).indexOf("\"") + 1);
+    }
+
+    private String prepareRole(String query) {
+        String[] values = query.split(":");
+        return values[2].substring(1, values[2].substring(1).indexOf("\"") + 1);
+    }
```
magic numbers

JeniaSan reviewed (approved):
Very good job! Amazing attention to details, I'm shocked

Someboty committed:
Removed magic numbers and some redundant variables

Someboty committed:
Removed more redundant variables

nacenik commented on src/main/resources/db/changelog/changes/05-create-users_roles-table.yaml:
```diff
@@ -0,0 +1,20 @@
+                type: bigint
+                constraints:
+                  foreignKeyName: role.id
+                  nullable: false
```
add empty line at the end of the file

nacenik commented on src/main/resources/db/changelog/changes/04-insert-roles.yaml:
```diff
@@ -0,0 +1,17 @@
+        - insert:
+            tableName: roles
+            columns:
+              - column: {name: role, value: "2"}
```
add empty line at the end of the file

nacenik commented on src/main/java/com/bookshop/controller/AuthController.java:
```diff
@@ -0,0 +1,122 @@
+            @ApiResponse(responseCode = "401",
+                    description = "User should be authenticated to do that operation"),
+            @ApiResponse(responseCode = "403",
+                    description = "Only users with role \"ADMIN\" can do such operation"),
```
403 could happen if the user tries to reach any endpoint without authorization

Someboty replied:
Agreed, but I tried to describe the situation in which this exception can occur. And it will be if users with roles USER or MANAGER try to use this endpoint. Should I rewrite this description?

nacenik commented on src/main/java/com/bookshop/controller/AuthController.java:
```diff
@@ -0,0 +1,122 @@
+            @ApiResponse(responseCode = "401",
+                    description = "User should be authenticated to do that operation"),
+            @ApiResponse(responseCode = "403",
+                    description = "Only users with role \"ADMIN\" can do such operation"),
```
the same

nacenik commented on src/main/java/com/bookshop/controller/AuthController.java:
```diff
@@ -0,0 +1,122 @@
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteById(@RequestBody String email) {
+        userService.deleteByEmail(email);
+    }
```
deleting users in the authorization controller????

nacenik commented on src/main/java/com/bookshop/controller/AuthController.java:
```diff
@@ -0,0 +1,122 @@
+    })
+    @Operation(summary = "set user's role",
+            description = "operation only for users with \"ADMIN\" role. works in both ways")
+    @PutMapping("/role")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    public UserRegistrationResponseDto setAsRole(
+            @RequestBody String query) throws RegistrationException {
+        return userService.setAsRole(query);
+    }
```
Any user can put admin authority???

Someboty replied:
Any user with role “ADMIN”

nacenik commented on src/main/java/com/bookshop/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,92 @@
+        return userMapper.toResponse(userRepository.save(user));
+    }
+
+    @Override
+    public UserRegistrationResponseDto setAsRole(String query) throws RegistrationException {
+        User user = findByEmail(prepareEmail(query));
+        String role = prepareRole(query);
+        switch (role) {
+            case ("admin") -> user.setRoles(Set.of(userRoleRepository
+                            .getUserRoleByName(UserRole.RoleName.ROLE_USER),
+                    userRoleRepository.getUserRoleByName(UserRole.RoleName.ROLE_MANAGER),
+                    userRoleRepository.getUserRoleByName(UserRole.RoleName.ROLE_ADMIN)));
+            case ("manager") -> user.setRoles(Set.of(userRoleRepository
+                            .getUserRoleByName(UserRole.RoleName.ROLE_USER),
+                    userRoleRepository.getUserRoleByName(UserRole.RoleName.ROLE_MANAGER)));
+            case ("user") -> user.setRoles(Set.of(userRoleRepository
+                    .getUserRoleByName(UserRole.RoleName.ROLE_USER)));
+            default -> throw new RegistrationException("Incorrect role: " + role);
+        }
+        return userMapper.toResponse(userRepository.save(user));
+    }
```
If I have registered on your website, could I put my role as admin or manager?

Someboty replied:
You can create only user account by your own. You need a user with ADMIN role to become MANAGER or ADMIN

Someboty replied:
Maybe I will add a method, where admin can create a user with some role (not create a user account and rank up it to some role separately)

nacenik commented on src/main/java/com/bookshop/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,92 @@
+        user.setPassword(passwordEncoder.encode(request.password()));
+        user.setFirstName(request.firstName());
+        user.setLastName(request.lastName());
+        user.setShippingAddress(request.shippingAddress());
```
use mapper

nacenik reviewed (changes requested):
Let's improve your solution.

Someboty committed:
Created UserController, moved deleting user and change user's role in UserController,
added method for creating user with some role (only for admin),
created dto for methods working with roles, added mapping for them,
 separated checking if user exists and setting roles in separate methods,
  fixed content problems with 401/403 responses, added examples for fields in CreateBookRequestDto,
  fixed style errors

vovasalyha commented on src/main/java/com/bookshop/controller/UserController.java:
```diff
@@ -0,0 +1,112 @@
+@RequiredArgsConstructor
+@RestController
+public class UserController {
+    private static final String INCORRECT_DATA =
```
You sure that you need to create this constant in each controller? Maybe it's better to have some centralised place for it

Someboty committed:
created class for messages, moved it with UserRole to new repository

Someboty committed:
fixed style mistakes
